### PR TITLE
Port RandAugment from pytorch

### PR DESCRIPTION
**Pull request by GilesBathgate**
I wanted to use RandAugment from the pytorch autoaugment module. https://pytorch.org/vision/main/generated/torchvision.transforms.RandAugment.html

I think this is line for line equivalent to the pytorch code, but maybe needs additional improvements specific to TorchSharp. It works as expected, but depends on my previous PRs

### CODE REVIEW

**Comment by NiklasGustafsson** [comment]
Three thoughts: 

1) Thank you very much for contributing! I'm impressed by the number of PRs that I returned to after my vacation. :-)

2) When you're adding new functionality, try to see if there's a unit test or two that you can add. When you fix a bug, add a unit test that would have failed with the bug, but passes after the fix.

3) When you add new functionality or fix a bug, add an entry to the RELEASENOTEs.md file, at the top under the next release heading. It makes it easier to highlight the changes when we make a release.

Thanks,

Niklas

**Comment by GilesBathgate** [comment]
@NiklasGustafsson Hope you had a nice vacation. I was holding off doing more until I had feedback on the PR, and also haven't found out how to add unit tests at this stage, so I could use some guidance. I implemented these changes for a work demonstration involving a multi-label classifier using vgg16. Since I needed RandAugment, for that demo, I had already implemented the code before contributing.

**Comment by NiklasGustafsson** [comment]
As long as you don't add new files, adding unit tests is very straight-forward. Just add a new `public void` function with the `[Fact]` attribute to the end of an existing file (try finding one with related tests) in the 'TestTorchSharp' project, call it something reasonable, and you're off to the races.

Also, for regression tests, the file `TestTensorBugs` is where I have collected most of them, even though they are not all about bugs in the `Tensor` code.

**Comment by NiklasGustafsson** [comment]
I wonder if it's as easy as checking (in `Tensor.Dispose()`) whether the associated `grad` Tensor is non-null and disposing it, too...

File: src/TorchVision/AutoAugment.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+using F = TorchSharp.torchvision.transforms.functional;
+
+#nullable enable
+namespace TorchSharp
+{
+    public static partial class torchvision
+    {
+        private static Tensor apply_op(
+                                Tensor img,
+                                string op_name,
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:14]: TorchSharp convention -- we use enums in situations like this.

File: src/TorchVision/AutoAugment.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                                float magnitude,
+                                InterpolationMode interpolation,
+                                IList<float>? fill)
+        {
+            if (op_name == "ShearX") {
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:19]: Once you have switched to enums, a switch statement may be faster and easier to read.

File: src/TorchVision/AutoAugment.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                // Pass
+            } else {
+                throw new ArgumentException($"The provided operator {op_name} is not recognized.");
+            }
+
+            return img;
+        }
+
+
+        internal class RandAugment : ITransform
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:85]: I would place the class in the 'Modules' namespace under 'TorchSharp,' which is the pattern we (have tried to) use elsewhere.

File: src/TorchVision/AutoAugment.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+using System.Linq;
+using static TorchSharp.torch;
+using F = TorchSharp.torchvision.transforms.functional;
+
+#nullable enable
+namespace TorchSharp
+{
+    public static partial class torchvision
+    {
+        private static Tensor apply_op(
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:12]: Should this just be a private method in the class? Where else is it used?

File: src/TorchVision/AutoAugment.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            } else if (op_name == "AutoContrast") {
+                img = F.autocontrast(img);
+            } else if (op_name == "Equalize") {
+                img = F.equalize(img);
+            } else if (op_name == "Invert") {
+                img = F.invert(img);
+            } else if (op_name == "Identity") {
+                // Pass
+            } else {
+                throw new ArgumentException($"The provided operator {op_name} is not recognized.");
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:78]: With an enum, you won't have to throw. Just sayin'... :-)

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]

File: src/TorchVision/AutoAugment.cs (hunk 5 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:78]: You don't need to convince me. I much prefer enums, or better still class based dispatch. 
the idea was to copy RandAugment, then update it: https://pytorch.org/vision/main/_modules/torchvision/transforms/autoaugment.html#_apply_op

File: src/TorchVision/AutoAugment.cs (hunk 4 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:12]: It will be used by `AutoAugment`, `TrivialAugmentWide`, etc...

Comment (NiklasGustafsson) [reply src/TorchVision/AutoAugment.cs:12]: Time for a base class with a protected member?

File: src/TorchVision/AutoAugment.cs (hunk 3 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs]: I don't think it needs to be exposed under the modules namespace, its `internal`. Other transforms see [Rotate.cs ](https://github.com/dotnet/TorchSharp/blob/main/src/TorchVision/Rotate.cs) follow the pattern I've used here.

Comment (NiklasGustafsson) [reply src/TorchVision/AutoAugment.cs]: It's easier to find when you're cross-referencing code in VS. 

Not a biggie, but consistency (if you find places where the current code is not) is helpful across a code base.

Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs]: @NiklasGustafsson I think all files under `TorchVision/*.cs` currently! I feel its a separate refactoring issue.

File: src/TorchVision/AutoAugment.cs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+                        return F.solarize(img, magnitude);
+                    case opType.AutoContrast:
+                        return F.autocontrast(img);
+                    case opType.Equalize:
+                        return F.equalize(img);
+                    case opType.Invert:
+                        return F.invert(img);
+                    case opType.Identity:
+                        return img; // Pass
+                    default:
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:89]: I'm surprised the C# compiler doesn't complain about the fact that you have already covered all the possible cases. The body of the default case is, in effect, unreachable code.

Comment (NiklasGustafsson) [reply src/TorchVision/AutoAugment.cs:89]: I take that back -- ShearY is not covered. Is that intentional?

File: src/TorchVision/AutoAugment.cs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                this.num_ops = num_ops;
+                this.magnitude = magnitude;
+                this.num_magnitude_bins = num_magnitude_bins;
+                this.interpolation = interpolation;
+                this.fill = fill;
+            }
+
+            public Tensor call(Tensor img)
+            {
+                var fill = this.fill;
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:113]: I would put a dispose scope declaration in this method, and move the result to the outer scope on return.

Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:113]: Why the local?

File: src/TorchVision/AutoAugment.cs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                var height = img.shape[cDim + 1];
+                var width = img.shape[cDim + 2];
+                var op_meta = augmentation_space(num_magnitude_bins, (height, width));
+                for (int i = 0; i < num_ops; ++i) {
+                    var op_index = torch.randint(op_meta.Count, new[] { 1 }).ToInt32();
+                    var op_name = op_meta.Keys.ToList()[op_index];
+                    var (magnitudes, signed) = op_meta[op_name];
+                    var magnitude = magnitudes.Dimensions > 0 ? magnitudes[this.magnitude] : 0.0f;
+
+                    if (signed && torch.randint(2, new[] { 1 }).ToBoolean())
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:124]: Is it possible to use `stackalloc` instead of `new` here? Probably not.

File: src/TorchVision/AutoAugment.cs (hunk 6 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:89]: hmm, No, did it never have, or did I remove it in the refactoring I wonder.

File: src/TorchVision/AutoAugment.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                        magnitude *= -1.0;
+
+                    img = apply_op(img, op_name, magnitude.ToSingle(), interpolation, fill);
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:127]: Since you have to have it as a float, anyway (should it be a double??), isn't it better to do the `ToSingle()` call in the true-branch of the conditional expression on line 122, and leave the local variable a float instead of a Tensor?

File: src/TorchVision/AutoAugment.cs (hunk 8 above)
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:133]: Hmmm... I wonder if there isn't a cheaper way for us to get a random scalar than this. I'll have to look into that.

Comment (NiklasGustafsson) [reply src/TorchVision/AutoAugment.cs:133]: Created https://github.com/dotnet/TorchSharp/issues/1068 to track

File: src/TorchVision/AutoAugment.cs (hunk 6 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:98]: @NiklasGustafsson Good catch, I've fixed that now. 

BTW I think c# allows you to cast any int to an enum even when that enum doesn't contain that value, maybe that was an old quirk that would emit a warning now.

File: src/TorchVision/AutoAugment.cs (hunk 7 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:122]: I've no idea. Pytorch has the same. will remove.

BTW I think its a good approach to copy as close to the original implementation line for line and then refactor, even if the original didn't make sense.

File: src/TorchVision/AutoAugment.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            private readonly int num_ops;
+            private readonly int magnitude;
+            private readonly int num_magnitude_bins;
+            private readonly InterpolationMode interpolation;
+            private readonly IList<float>? fill;
+        }
+
+        public static partial class transforms
+        {
+            static public ITransform RandAugment(
```
Comment (NiklasGustafsson) [inline src/TorchVision/AutoAugment.cs:170]: Doc comments, of course.

File: src/TorchVision/AutoAugment.cs (hunk 7 above)
Comment (NiklasGustafsson) [reply src/TorchVision/AutoAugment.cs:122]: Definitely. I do the same.

File: src/TorchVision/AutoAugment.cs (hunk 8 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:133]: Ah, good call. Never came across `stackalloc` before it must be because I am :older_man:

File: src/TorchVision/AutoAugment.cs (hunk 7 above)
Comment (NiklasGustafsson) [reply src/TorchVision/AutoAugment.cs:122]: Oh -- need to say this: for anything with any kind of algorithm in it, I'm always try to be careful to add a link to the original Python file at the top of the file, for attribution as well as reference. 

See TorchVision/Ops/Permute.cs, for example.

File: src/TorchVision/AutoAugment.cs (hunk 9 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:136]: Yeah, I'll make it double too.

File: src/TorchVision/AutoAugment.cs (hunk 8 above)
Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs]: Just the wrong parameters here. Fixed.

**Comment by NiklasGustafsson** [comment]
@GilesBathgate. All my comments have been addressed, except the need for some unit tests.

**Comment by GilesBathgate** [comment]
> @GilesBathgate. All my comments have been addressed, except the need for some unit tests.

Yep. Will make some progress on those over the w/end.

**Comment by NiklasGustafsson** [comment]
> > @GilesBathgate. All my comments have been addressed, except the need for some unit tests.
> 
> Yep. Will make some progress on those over the w/end.

Great. While you're at it, add some text at the beginning of the file to point at the Python source code.

Also -- every time you push, a new build starts, but the last one keeps going (it looks like it's been canceled, but it hasn't). It will save some compute resources if you batch up a number of commits into one push.

**Comment by NiklasGustafsson** [comment]
And add a line for this PR and the make_grid PR in the RELASENOTES.md file, under the 0.100.4 release notes at the top.

**Comment by GilesBathgate** [comment]
@NiklasGustafsson If the you think the unit tests are sufficient, then this PR is completed.

**Comment by NiklasGustafsson** [comment]
> @NiklasGustafsson If the you think the unit tests are sufficient, then this PR is completed.

Yup. I'll try to get a clean build, then merge.

**Review by NiklasGustafsson — APPROVED** [review]

File: src/TorchVision/AutoAugment.cs [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                return img.MoveToOuterDisposeScope();
+            }
+
+            private Dictionary<opType, (Tensor, bool)> augmentation_space(int num_bins, (long height, long width) image_size)
+            {
+                return new Dictionary<opType, (Tensor, bool)> {
+                    { opType.Identity, (torch.tensor(0.0), false) },
+                    { opType.ShearX, (torch.linspace(0.0, 0.3, num_bins), true) },
+                    { opType.ShearY, (torch.linspace(0.0, 0.3, num_bins), true) },
+                    { opType.TranslateX, (torch.linspace(0.0, 150.0 / 331.0 * image_size.height, num_bins), true) },
```
Comment (GilesBathgate) [inline src/TorchVision/AutoAugment.cs:149]: @NiklasGustafsson I think I spotted a bug here, TranslateX should not be image_size.height, it should be image_size.width, TranslateY vice versa.

Comment (GilesBathgate) [reply src/TorchVision/AutoAugment.cs:149]: Fixed as part of #1077
